### ISSUE TITLE
[scala] Accuracy precision is too low

### ISSUE THREAD

**Issue by benqua** [apache/incubator-mxnet#8226]
## Description
When using the Accuracy EvalMetric, it is very easy to fall in cases where `EvalMetric.sumMetric` has a too low precision.

## Environment info (Required)
ubuntu, scala (not relevant).

## Build info (Required if built from source)
MXNet commit hash: 0449cc646a9f5d76e6c9fb01659e771bf1237233

## Description
In EvalMetric.scala, sumMetric is defined as:
```scala
  protected var sumMetric: Float = 0.0f
```
and the Accuracy metric does:
```scala
      for ((labelElem, predElem) <- label.toArray zip predLabel.toArray) {
        if (labelElem == predElem) {
          this.sumMetric += 1
        }
      }
```
Because sumMetric is a float, if it becomes bigger than ~10^7, then `this.sumMetric += 1` will have no effect and sumMetric will not be updated anymore. It is quite easy to get big numbers in sumMetric while solving, for example, image segmentation tasks (with output size of ~ 800x800, it will take about 50 output to get to 10^7). This won't throw any warning or error.

## Solutions
I worked around this issue by

1. summing all elements of one output before adding the result to sumMetric
2. dividing the sum by the number of elements

### 1) summing element before adding the result to sumMetric
To lower the difference of magnitude between this.sumMetric and 1, i replaced
```scala
      for ((labelElem, predElem) <- label.toArray zip predLabel.toArray) {
        if (labelElem == predElem) {
          this.sumMetric += 1
        }
      }
```
by
```scala
val ok = labe.toArrayl.zip(predLabel.toArray)
           .filter{ case (lElem: Float, pElem: Float) => lElem == pElem }.size.toFloat
this.sumMetric += s
```
s being (in my case) far bigger than 1, this change push back the problem. However, it doesn't solve it and that was not enough in my case.

### 2) Divide the sum by the number of elements
It would be better if the Accuracy doesn't depend on the size of the output. Hence, I divided the sum by the number of elements
 ```scala
val la = label.toArray
val size = label.size.toFloat
val ok = la.zip(predLabel.toArray)
            .filter{ case (lElem: Float, pElem: Float) => lElem == pElem }.size.toFloat
this.sumMetric += ok / size
```
With this change, I was able to train my network with about 50k inputs.

## Additional better (?) solution
We could use a `double` for this.sumMetric of EvalMetric, which would likely be enough for most case. The `get` function could still return a casted float if the EvalMetric API cannot easily be changed. 

I can provide a PR for the fix 1) and 2).
I am waiting for comments and feedbacks regarding the use of double for sumMetric.

My PR would also fix #8129 if @Ldpe2G (who found the solution) did not do it in the meantime. 

(By the way, am I the only one using Accuracy this way in Scala?)

**Comment by Ldpe2G** [comment]
@benqua looking forward to your pr, and pls remember to test the metric with both segmentation and classification task. By the way I think you should also use the mean IOU metric for your segmentation task.

**Comment by benqua** [comment]
@Ldpe2G I did a PR:
https://github.com/apache/incubator-mxnet/pull/8297
I also did another PR that fix #8129 following your advice.

@Ldpe2G @javelinjs can you please comment and review?
